ethers-providers/src/transports/ipc.rs

```diff
@@ -1,48 +1,217 @@
     sync::{
         atomic::{AtomicU64, Ordering},
         Arc,
     },
     thread,
 };
-
-use async_trait::async_trait;
-use bytes::{Buf as _, BytesMut};
-use ethers_core::types::U256;
-use futures_channel::mpsc;
-use futures_util::stream::StreamExt as _;
-use hashers::fx_hash::FxHasher64;
-use serde::{de::DeserializeOwned, Serialize};
-use serde_json::{value::RawValue, Deserializer};
 use thiserror::Error;
 use tokio::{
-    io::{AsyncReadExt as _, AsyncWriteExt as _, BufReader},
-    net::{
-        unix::{ReadHalf, WriteHalf},
-        UnixStream,
-    },
+    io::{AsyncReadExt, AsyncWriteExt, BufReader},
     runtime,
     sync::oneshot::{self, error::RecvError},
 };
 
-use crate::{
-    provider::ProviderError,
-    transports::common::{JsonRpcError, Request, Response},
-    JsonRpcClient, PubsubClient,
-};
-
-use super::common::Params;
-
 type FxHashMap<K, V> = std::collections::HashMap<K, V, BuildHasherDefault<FxHasher64>>;
 
 type Pending = oneshot::Sender<Result<Box<RawValue>, JsonRpcError>>;
 type Subscription = mpsc::UnboundedSender<Box<RawValue>>;
 
-/// Unix Domain Sockets (IPC) transport.
+#[cfg(unix)]
+#[doc(hidden)]
+mod imp {
+    pub(super) use tokio::net::{
+        unix::{ReadHalf, WriteHalf},
+        UnixStream as Stream,
+    };
+}
+
+#[cfg(windows)]
+#[doc(hidden)]
+mod imp {
+    use super::*;
+    use std::{
+        ops::{Deref, DerefMut},
+        pin::Pin,
+        task::{Context, Poll},
+        time::Duration,
+    };
+    use tokio::{
+        io::{AsyncRead, AsyncWrite, ReadBuf},
+        net::windows::named_pipe::{ClientOptions, NamedPipeClient},
+        time::sleep,
+    };
+    use winapi::shared::winerror;
+
+    /// Wrapper around [NamedPipeClient] to have the same methods as a UnixStream.
+    ///
+    /// Should not be exported.
+    #[repr(transparent)]
+    pub(super) struct Stream(pub NamedPipeClient);
+
+    impl Deref for Stream {
+        type Target = NamedPipeClient;
+
+        fn deref(&self) -> &Self::Target {
+            &self.0
+        }
+    }
+
+    impl DerefMut for Stream {
+        fn deref_mut(&mut self) -> &mut Self::Target {
+            &mut self.0
+        }
+    }
+
+    impl Stream {
+        pub async fn connect(addr: impl AsRef<Path>) -> Result<Self, io::Error> {
+            let addr = addr.as_ref().as_os_str();
+            loop {
+                match ClientOptions::new().open(addr) {
+                    Ok(client) => break Ok(Self(client)),
+                    Err(e) if e.raw_os_error() == Some(winerror::ERROR_PIPE_BUSY as i32) => (),
+                    Err(e) => break Err(e),
+                }
+
+                sleep(Duration::from_millis(50)).await;
+            }
+        }
+
+        #[allow(unsafe_code)]
+        pub fn split(&mut self) -> (ReadHalf, WriteHalf) {
+            // SAFETY: ReadHalf cannot write but still needs a mutable reference for polling.
+            // NamedPipeClient calls its `io` using immutable references, but it's private.
```

Comment:
I believe this is correct, there can be two tasks that use the underlying `io` as long as one only reads and the other only writes.

`UnixStream` basically does the same but implements it slightly differently:

it uses immutable references in the halves

https://docs.rs/tokio/latest/src/tokio/net/unix/split.rs.html#51-53

and uses unsafe in separate poll functions that use unsafe

https://docs.rs/tokio/latest/src/tokio/net/unix/stream.rs.html#944-950

perhaps we mimick this as well?

Reply:
the problem is the read/write trait poll methods need a mutable reference because self is `Pin<&mut self>` so that's why I'm storing a mut ref instead of immutable. Otherwise if the inner `io` was public I'd be calling that since it uses immutable references

Reply:
makes sense